Update the outdated Docker guide (#229)
crate/crate-howtos: docs/deployment/containers/docker.rst

  to allocate 2G for its heap
- Runs the command ``crate`` inside the container with parameters:

>     * ``network.host``: The ``_site_`` value results in the binding of the

Reviewer: I think it is better to do lowercase after a colon throughout, unless it would really be confusing to do so.

      CrateDB process to a site-local IP address.
    * ``node.name``:  Defines the node's name as ``crate01`` (used by
      master election).
    * ``discovery.seed_hosts``: This parameter lists the other hosts in the
      cluster. The format is a comma-separated list of ``host:port`` entries,
      where port defaults to setting ``transport.tcp.port``. Each node must
      contain the name of all the other hosts in this list. Notice also that
      any node in the cluster might be started at any time, and this will
      create connection exceptions in the log files, however all nodes will
      eventually be running and interconnected.
    * ``cluster.initial_master_nodes``: Defines the list of master-eligible
      node names which will participate in the vote of the first master
      (first bootstrap). If this parameter is not defined, then it is expected
      that the node will join an already formed cluster. This parameter is only
      relevant for the first election.

>     * ``gateway.expected_nodes`` and ``gateway.recover_after_nodes``: Specifies
>       how many nodes you want in the cluster and how the cluster's state should
>       be recovered once all nodes are started.

Reviewer 2: ```suggestion
    * ``gateway.expected_nodes`` and ``gateway.recover_after_nodes``: Specifies
      how many nodes you expect in the cluster and how many nodes must be discovered before the cluster state is recovered.
```

Notice here that:

> - You updated the container and node name to ``crate02``

Reviewer: Would these lists, including also further below, benefit from semicolons at the end (with final period)? Seems a bit more readable to me (but up to you)

                  "-Cnode.name=cratedb02",
                  "-Cnode.data=true",
                  "-Cdiscovery.seed_hosts=cratedb01,cratedb03",
                  "-Ccluster.initial_master_nodes=cratedb01,cratedb02,cratedb03",
                  "-Cgateway.expected_nodes=3",
                  "-Cgateway.recover_after_nodes=2"]  
        deploy:
          replicas: 1
          restart_policy:
            condition: on-failure
        environment:
          - CRATE_HEAP_SIZE=2g
      
      cratedb03:
        image: crate:latest    
        ports:
          - "4203:4200"
        volumes:
          - /tmp/crate/03:/data
        command: ["crate",
                  "-Ccluster.name=crate-docker-cluster",
                  "-Cnode.name=cratedb03",
                  "-Cnode.data=true",
                  "-Cdiscovery.seed_hosts=cratedb01,cratedb02",
                  "-Ccluster.initial_master_nodes=cratedb01,cratedb02,cratedb03",
                  "-Cgateway.expected_nodes=3",
                  "-Cgateway.recover_after_nodes=2"]
        deploy:
          replicas: 1
          restart_policy:
            condition: on-failure
        environment:
          - CRATE_HEAP_SIZE=2g

> In file above:

Reviewer: In the file above

this reason, you should mount a persistent ``data`` directory on your host
machine to the ``/data`` directory inside the container::

    $ docker run -d -v /srv/crate/data:/data crate \
        crate -Cnetwork.host=_site_

Here, ``/srv/crate/data`` is an example path, and should be replaced with the
path to your host machine's ``data`` directory.

See the `Docker volume`_ documentation for more help.

Custom configuration
--------------------

If you want to use a custom configuration, it is recommended that you mount
configuration files on the host machine to the appropriate path inside the
container. That way, your configuration will not be lost if the container is
removed.

Here is an example of how you could mount the ``crate.yml`` config file::

    $ docker run -d \
        -v /srv/crate/config/crate.yml:/crate/config/crate.yml crate \
        crate -Cnetwork.host=_site_

Here, ``/srv/crate/config/crate.yml`` is an example path, and should be
replaced with the path to your host machine's ``crate.yml`` file.

Troubleshooting
===============

The official `CrateDB Docker image`_ ships with a liveness `healthcheck`_
configured.

> This healthcheck will flag a problem if the CrateDB process crashed or hung

Reviewer: "Health check" imho. Don't think it's conventional enough to have a compound without space.

Resource constraints
====================

It is important that you set resource constraints when you are running CrateDB

> inside Docker. Current versions of the JVM are unable to detect that they are

Reviewer 2: This is a bit outdated. This was fixed at some point with JDK 8, and hasn't been a problem with JDK 11 (which we require since CrateDB 4.0).

I think we could remove this entire paragraph. Or maybe point out that users should work with cpu/memory limits if they plan to run multiple CrateDB containers on a single machine, to avoid CrateDB from over allocating resources

Resource constraints
====================

To avoid overallocation of resources, you may want to consider setting
constraints on CPU and memory if you plan to run multiple CrateDB containers
on a single machine.

Bootstrap checks
----------------

When using CrateDB with Docker, CrateDB binds by default to any site-local IP
address on the system (i.e. 192.168.0.1). This performs a number of checks 
during bootstrap. The settings listed in `Bootstrap Checks`_ must be addressed on
the Docker **host system** in order to start CrateDB successfully and when 
`going into production`_.

Memory
------

You must calculate and explicitly `set the maximum memory`_ that the container

> can use. This is dependant on your host system and should typically be as high

Reviewer: Dependent

as possible.

You must then calculate the appropriate heap size (typically half the container's
memory limit, see `CRATE_HEAP_SIZE`_ for details) and pass this to CrateDB,
which in turn passes it to the JVM.

It is not necessary to configure swap memory since CrateDB does not use swap.

CPU
---

You must calculate and explicitly `set the maximum number of CPUs`_ that the

> container can use. This is dependant on your host system and should typically

Reviewer: Dependent here also
